### PR TITLE
Matrix Service for M4974 (diagnostics platform app)

### PR DESCRIPTION
**Pull request by bartcharbon**
#### Checklist
- [ ] Functionality works & meets specifications
- [ ] Code reviewed
- [ ] Code unit/integration/system tested
- [ ] User documentation updated
- [ ] (If you have changed REST API interface) view-swagger.ftl updated
- [ ] Test plan template updated [MatrixTestPlan.xlsx](https://github.com/molgenis/molgenis/files/941187/MatrixTestPlan.xlsx)
- [ ] Clean commits

### CODE REVIEW

**Comment by bartcharbon** [comment]
test this please

File: molgenis-data-matrix/src/main/java/org/molgenis/data/matrix/impl/DoubleMatrix.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+            setRowIndicesMap();
+            setColumnIndicesMap();
+        } catch (IOException e) {
+            throw new MolgenisDataException(e);
+        }
+        inited = true;
+    }
+
+    public double getValueByIndex(int row, int column) {
+        if (!inited) init();
```
Comment (Mark-de-Haan) [inline molgenis-data-matrix/src/main/java/org/molgenis/data/matrix/impl/DoubleMatrix.java]: Call init() from constructor?

File: molgenis-security/src/main/java/org/molgenis/security/MolgenisWebAppSecurityConfig.java [hunk 2]
```diff
@@ -148,6 +148,8 @@ protected void configure(HttpSecurity http) throws Exception
 
 				.antMatchers("/api/**").permitAll()
 
+				.antMatchers("/matrix/**").permitAll()
```
Comment (Mark-de-Haan) [inline molgenis-security/src/main/java/org/molgenis/security/MolgenisWebAppSecurityConfig.java]: is this necessary when matrix is a sub namespace of api?

File: molgenis-data-matrix/src/main/java/org/molgenis/data/matrix/MatrixMapper.java [hunk 3]
```diff
@@ -0,0 +1,5 @@
+package org.molgenis.data.matrix;
+
+public interface MatrixMapper {
```
Comment (Mark-de-Haan) [inline molgenis-data-matrix/src/main/java/org/molgenis/data/matrix/MatrixMapper.java]: Did you run formatter on java files?
My formatter changes everything...

File: molgenis-data-matrix/src/main/java/org/molgenis/data/matrix/impl/DoubleMatrix.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+            throw new IndexOutOfBoundsException("Index [" + column + "] is greater than the number of columns in the matrix [" + matrix.getColumnCount() + "]");
+        return matrix.getAsDouble(row, column);
+    }
+
+    public double getValueByName(String row, String column) {
+        if (!inited) init();
+        Integer rowIndex = rowMap.get(row);
+        Integer columnIndex = columnMap.get(column);
+        if (rowIndex == null) throw new MolgenisDataException("Unknown row name [" + row + "]");
+        if (rowIndex == null) throw new MolgenisDataException("Unknown column name [" + column + "]");
```
Comment (Mark-de-Haan) [inline molgenis-data-matrix/src/main/java/org/molgenis/data/matrix/impl/DoubleMatrix.java]: Shouldnt this be columnIndex == null?

File: molgenis-data-matrix/src/main/java/org/molgenis/data/matrix/impl/MatrixServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    private Map<String, MatrixMapper> rowMappingMap = new HashMap<>();
+
+    @Autowired
+    public MatrixServiceImpl(DataService dataService, FileStore fileStore) {
+        this.dataService = dataService;
+        this.fileStore = fileStore;
+    }
+
+    private DoubleMatrix getMatrixByEntityTypeId(String entityId) {
+        DoubleMatrix matrix = null;
```
Comment (Mark-de-Haan) [inline molgenis-data-matrix/src/main/java/org/molgenis/data/matrix/impl/MatrixServiceImpl.java]: Variable is never used

File: molgenis-data-matrix/src/main/java/org/molgenis/data/matrix/impl/MatrixServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+        } else {
+            throw new MolgenisDataException("Unknown Matrix Metadata EntityID [" + entityId + "]");
+        }
+    }
+
+    private DoubleMatrix getMatrix(String entityId, Entity entity) {
+        DoubleMatrix matrix;
+        if (matrixMap.get(entityId) == null) {
+            String fileLocation = entity.getString(MatrixMetadata.FILE_LOCATION);
+            char seperator = MatrixMetadata.getSeparatorValue(entity.getString(MatrixMetadata.SEPERATOR));
```
Comment (Mark-de-Haan) [inline molgenis-data-matrix/src/main/java/org/molgenis/data/matrix/impl/MatrixServiceImpl.java]: separator*

File: molgenis-data-matrix/src/main/java/org/molgenis/data/matrix/impl/MatrixServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+    }
+
+    private void createMapper(String entityId, FileMeta meta, Map<String, MatrixMapper> mapperMap) {
+        File file = fileStore.getFile(meta.getId());
+        MatrixMapper mapper = new MatrixMapperImpl(file);
+        mapperMap.put(entityId, mapper);
+    }
+
+    @Override
+    @RequestMapping(value = "{entityId}/valueByIndex", method = GET)
```
Comment (Mark-de-Haan) [inline molgenis-data-matrix/src/main/java/org/molgenis/data/matrix/impl/MatrixServiceImpl.java]: Request mapping values start with /

File: molgenis-data-matrix/src/main/java/org/molgenis/data/matrix/impl/MatrixServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+    @ResponseBody
+    public Object getValueByIndex(@PathVariable("entityId") String entityName, @RequestParam("row") int row, @RequestParam("column") int column) throws MolgenisDataException {
+        DoubleMatrix matrix = getMatrixByEntityTypeId(entityName);
+        return matrix.getValueByIndex(row, column);
+    }
+
+    @Override
+    @RequestMapping(value = "{entityId}/valueByNames", method = GET)
```
Comment (Mark-de-Haan) [inline molgenis-data-matrix/src/main/java/org/molgenis/data/matrix/impl/MatrixServiceImpl.java]: /{entityId}/valueByNames*

File: molgenis-data-matrix/src/main/java/org/molgenis/data/matrix/model/Score.java [hunk 9]
```diff
@@ -0,0 +1,57 @@
+package org.molgenis.data.matrix.model;
+
+public class Score {
```
Comment (Mark-de-Haan) [inline molgenis-data-matrix/src/main/java/org/molgenis/data/matrix/model/Score.java]: You can let AutoValue do the work

```
@AutoValue
@AutoGson(autoValueClass = AutoValue_Score.class)
public abstract class Score
{
	public abstract String getColumn();

	public abstract String getRow();

	public abstract Object getValue();

	public static Score createScore(String column, String row, Object value)
	{
		return new AutoValue_Score(column, row, value);
	}
}
```

File: molgenis-data-matrix/src/main/java/org/molgenis/data/matrix/MatrixMapper.java (hunk 3 above)
Comment (bartcharbon) [inline molgenis-data-matrix/src/main/java/org/molgenis/data/matrix/MatrixMapper.java]: fixed this once, however somehow intellij reverted to default... will fix...

File: molgenis-data-matrix/src/main/java/org/molgenis/data/matrix/impl/MatrixServiceImpl.java (hunk 7 above)
Comment (bartcharbon) [inline molgenis-data-matrix/src/main/java/org/molgenis/data/matrix/impl/MatrixServiceImpl.java]: works fine this way, but I'll add a slash

File: molgenis-data-matrix/src/main/java/org/molgenis/data/matrix/impl/MatrixServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+        File file = fileStore.getFile(meta.getId());
+        MatrixMapper mapper = new MatrixMapperImpl(file);
+        mapperMap.put(entityId, mapper);
+    }
+
+    @Override
+    @RequestMapping(value = "{entityId}/valueByIndex", method = GET)
+    @ResponseBody
+    public Object getValueByIndex(@PathVariable("entityId") String entityName, @RequestParam("row") int row, @RequestParam("column") int column) throws MolgenisDataException {
+        DoubleMatrix matrix = getMatrixByEntityTypeId(entityName);
```
Comment (Mark-de-Haan) [inline molgenis-data-matrix/src/main/java/org/molgenis/data/matrix/impl/MatrixServiceImpl.java]: Can matrix be null here?

File: molgenis-data-matrix/src/main/java/org/molgenis/data/matrix/impl/MatrixServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+    private DataService dataService;
+    private FileStore fileStore;
+
+    private Map<String, DoubleMatrix> matrixMap = new HashMap<>();
+    private Map<String, MatrixMapper> columnMappingMap = new HashMap<>();
+    private Map<String, MatrixMapper> rowMappingMap = new HashMap<>();
+
+    @Autowired
+    public MatrixServiceImpl(DataService dataService, FileStore fileStore) {
+        this.dataService = dataService;
```
Comment (Mark-de-Haan) [inline molgenis-data-matrix/src/main/java/org/molgenis/data/matrix/impl/MatrixServiceImpl.java]: Use Objects.requireNonNull()

**Comment by bartcharbon** [comment]
The code was moved to the molgenis-ui repository

**Pull request closed without merging**